### PR TITLE
Automatic creation of screenshots for user documentation

### PR DESCRIPTION
**Pull request by 3nids**
The idea is in the title.
This a first try, code is not ready yet, but it's open for comments on the structure of the code.

It's in app, so anyone who breaks a screenshot will be able to fix it!
That will make screenshots easily localized and/or OS specific.

One can run QGIS with `--take-screenshots __YOUR_PATH__`

### CODE REVIEW

File: src/app/qgisapp.h [hunk 1]
```diff
@@ -696,6 +696,9 @@ class APP_EXPORT QgisApp : public QMainWindow, private Ui::MainWindow
     //! Returns pointer to the identify map tool - used by identify tool in 3D view
     QgsMapToolIdentifyAction *identifyMapTool() const { return mMapTools.mIdentify; }
 
+    //! Take screenshots for user documentation
```
Comment (nyalldawson) [inline src/app/qgisapp.h]: What's "categories" mean? More dox required here

File: src/app/qgsappscreenshots.cpp [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+#include "qgsvectorlayerproperties.h"
+#include "qgsvectorlayer.h"
+#include "qgsproject.h"
+#include "qgsmessagelog.h"
+#include "qgisapp.h"
+
+QgsAppScreenShots::QgsAppScreenShots( const QString &saveDirectory )
+  : mSaveDirectory( saveDirectory )
+{
+  QString layerDef = QStringLiteral( "Point?crs=epsg:4326&field=pk:integer&field=my_text:string&field=my_integer:integer&field=my_double:double&key=pk" );
```
Comment (nyalldawson) [inline src/app/qgsappscreenshots.cpp]: We probably want some standardised, beautifully styled layers here?

File: src/app/qgsappscreenshots.cpp [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+      // TODO
+      return;
+  }
+
+  switch ( corner )
+  {
+    case Qt::BottomLeftCorner:
+      widget->move( 0, screenGeom.height() - widget->frameGeometry().height() );
+      break;
+    case Qt::BottomRightCorner:
```
Comment (nyalldawson) [inline src/app/qgsappscreenshots.cpp]: are these TODOs?

File: src/app/qgisapp.cpp [hunk 4]
```diff
@@ -13061,6 +13062,12 @@ void QgisApp::zoomToBookmarkIndex( const QModelIndex &index )
   mBookMarksDockWidget->zoomToBookmarkIndex( index );
 }
 
+void QgisApp::takeAppScreenShots( const QString &saveDirectory, const int categories )
+{
+  QgsAppScreenShots ass( saveDirectory );
+  ass.takeScreenshots( QgsAppScreenShots::Categories( categories ) );
```
Comment (nyalldawson) [inline src/app/qgisapp.cpp]: Can you rename "takeScreenshots" to "takePicturesOf" ? ;)

**Comment by nyalldawson** [comment]
@luipir @elpaso @volaya Didn't boundless have something similar? Possibly we could take advantage of some of their code/approaches?

**Comment by luipir** [comment]
@nyalldawson are you referring probably to lessons plugin? @SrNetoChan is more involved in document and autometic document creation and can help more about this part

**Comment by SrNetoChan** [comment]
@3nids thank you for this work! This was on my non-coder mind along time ago. It would help a lot in updating the screenshots for the Users documentation.

@nyalldawson the closest thing Boundless has is a small "plugin" to help creating lessons for the lessons plugin. This was Victor's work, but we ended up not using it much. If I recall, what it would do was to take a picture everytime there was a changes of focus in the dialogs. Here's the code:

https://github.com/boundlessgeo/qgis-lessons-plugin/tree/master/lessonscreator

File: src/app/qgsappscreenshots.cpp (hunk 3 above)
Comment (3nids) [reply src/app/qgsappscreenshots.cpp]: yes, they are...not sure we'll need them

File: src/app/qgsappscreenshots.cpp (hunk 2 above)
Comment (3nids) [reply src/app/qgsappscreenshots.cpp]: that would be _wunderbar_

File: src/app/qgisapp.cpp (hunk 4 above)
Comment (3nids) [reply src/app/qgisapp.cpp]: fixed

**Comment by DelazJ** [comment]
Thanks @3nids for pushing this forward.
About the areas to cover prioritarily, i'd say Layer properties, Project properties, Options dialog (not all at the beginning and in this order but these are the most changing and impacting dialogs afaict)

About lessons plugin, there was also @alexbruy [call](http://osgeo-org.1560.x6.nabble.com/QEP-intregrating-Lessons-plugin-into-QGIS-td5309638.html) and [qep](https://github.com/qgis/QGIS-Enhancement-Proposals/issues/92) to integrate it in core but nobody seems unfortunately interested. Based on the plugin's docs, I think this would have been a good addition to QGIS, and particularly to update training manual.
 
@SrNetoChan looks like the plugin has been ported to QGIS3 ?

**Comment by nyalldawson** [comment]
Got a screenshot of the gradient effect?

**Comment by SrNetoChan** [comment]
@DelazJ yes the lessons plugin was ported to qgis3, but lessons creator didn't.

File: src/app/qgisapp.cpp [hunk 5]
```diff
@@ -13061,6 +13062,12 @@ void QgisApp::zoomToBookmarkIndex( const QModelIndex &index )
   mBookMarksDockWidget->zoomToBookmarkIndex( index );
 }
 
+void QgisApp::takeAppScreenShots( const QString &saveDirectory, const int categories )
+{
+  QgsAppScreenShots ass( saveDirectory );
```
Comment (nyalldawson) [inline src/app/qgisapp.cpp]: For maximum LOLs, I think you should utilise the old school "my" prefix here.

Comment (3nids) [reply src/app/qgisapp.cpp]: I've thought about it, but sadly, this goes against our guidelines.

**Comment by 3nids** [comment]
@nyalldawson  here is the gradient:

25dsymbol

**Comment by 3nids** [comment]
Now this creates screenshots for both vector layer and global options.

I am running out of the initial time and would like to get feedback.

You can run `QGIS_EXE --take_screenshots DIRECTORY` and this will take and save the screenshots. They are already saved in a tree structure compatible with the documentation (`DIRECTORY` could be
`QGIS-Documentation/source/docs/user_manual`).

@DelazJ @ghtmtt @yjacolin what do you think? Would you like to merge this so you can test or you can build this branch to test? (to me, this is in a mergeable state, not perfect, but mergeable ;) )

**Comment by DelazJ** [comment]
Thanks @3nids 
i'm stuck to Windows and can't test other than osgeo4w, unless you have a custom build (i remember @elpaso provided some self-contained zip file in the past to test fixes on Windows)
About the screenshots, what's the output resolution. In docs, i think we should use 135dpi (see https://docs.qgis.org/testing/en/docs/documentation_guidelines/writing.html#add-new-screenshots). Is that the case?
i also wonder how the size of the screenshot is set. Could be nice to avoid big dialog with a lot of blank space at the bottom of the options (see eg in introduction folder the advanced -though i'm not sure we'll use this one-, authentication and canvas_and_legend screenshots). 
Out of curiosity, how is the image file name chosen?

**Comment by 3nids** [comment]
For the output resolution, is this something that could be done on Qt side? I just added the DPR to the pixmap to avoid doubling the size of the screenshot. But I guess it's not the way to define the output DPI.

File names are either defined manually in the code (like style, 2.5D symbol) or automatically while looping along the pages of the options based dialogs:

    QString name = dlg->mOptionsListWidget->item( row )[0].text().toLower();
    name.replace( " ", "_" ).replace( "&", "and" );

Of course, everything is possible there, we could create a map with our own names or define them in the UI or Cpp class.

For the size, we can easily define the dialog physical size and/or the screenshot area. I am thinking of creating a configuration file to avoid hardcoding these values (no need to recompile, easy definition of OS-based config file).

I'd propose to merge this one during the weekend so you can give it a try next week.

File: src/gui/qgisinterface.h [hunk 6]
```diff
@@ -528,6 +528,14 @@ class GUI_EXPORT QgisInterface : public QObject
      */
     virtual QgsStatusBar *statusBarIface() = 0;
 
+    /**
+     * Take screenshots for user documentation
+     * @param saveDirectory path were the screenshots will be saved
```
Comment (DelazJ) [inline src/gui/qgisinterface.h]: where

**Comment by DelazJ** [comment]
> I'd propose to merge this one during the weekend so you can give it a try next week.

👍

**Comment by ghtmtt** [comment]
I already sent an email but maybe it is worth to add also here my thoughts after the build. I'm running QGIS on Debian Sid.

Some problems with:

1. windows size: seems not correctly cut at the borders:

actions
 
2. names of the screenshots vs what they contain. The following screenshot's name is `auxiliary-storage.png` while it is actually the join tab, it happens for many of them:

auxiliary_storage

**Comment by elpaso** [comment]
> Thanks @3nids
> i'm stuck to Windows and can't test other than osgeo4w, unless you have a custom build (i remember @elpaso provided some self-contained zip file in the past to test fixes on Windows)

Unfortunately you cannot use them anymore (build is broken) since when the native implementation of desktop notifications has been changed to require some windows specific headers :(

**Comment by ghtmtt** [comment]
> alright, let's discuss here, it's a nice issue tracker :)
> 
>     1. Are you using 2 screens? I have observed weird behavior with 2 screens (with different resolution).
> 
>     2. I don't confirm this one. This is quite strange.

Nope, single screen on the laptop

**Comment by 3nids** [comment]
@ghtmtt ok, we'll try to tackle this one when we agree on doing something...the road still seems a bit complicate

@SrNetoChan tnanks for your feedback. I thought it was not possible to do that at first in Python...it could be. We could make the code returns dialogs and you coud access elements using findChildren.
I have to think a bit more at the code and will propose again here.

**Comment by SrNetoChan** [comment]
@3nids see if this piece of code, written by @luipir helps in any way.

https://github.com/boundlessgeo/qgis-lessons-plugin/blob/master/lessons/utils.py#L260

**Comment by DelazJ** [comment]
Trying with osgeo4w, I got dialogs less complete than @ghtmtt 's, eg:

joins

**Comment by DelazJ** [comment]
I also get crash at the end of the process
> Crash ID: db1d62a04d7bb6196d7b089042e9fcc842049ed2
>
> Stack Trace
> 
> QgsExpression::ParserError::ParserError :
> QgsOffsetUserWidget::offsetChanged :
> QgsOffsetUserWidget::offsetChanged :
> QgsOffsetUserWidget::offsetChanged :
> QgsOffsetUserWidget::offsetChanged :
> QgsOffsetUserWidget::offsetChanged :
> QgsOffsetUserWidget::offsetChanged :
> QgsOffsetUserWidget::offsetChanged :
> BaseThreadInitThunk :
> RtlUserThreadStart :

**Pull request merged**